[FLINK-3800] [runtime] Introduce SUSPENDED job status

tillrohrmann opened the pull request:
The SUSPENDED job status is a new ExecutionGraph state which can be reached from all
non-terminal states when calling suspend on the ExecutionGraph. Unlike the FAILED,
FINISHED and CANCELED state, the SUSPENDED state does not trigger the deletion of the
job from the HA storage. Therefore, this state can be used to handle the loss of
leadership or the shutdown of a JobManager so that the ExecutionGraph is stopped but
can still be recovered. SUSPENDED is also a terminal state but it can be differentiated as
a locally terminal state from FAILED, CANCELED and FINISHED which are globally
terminal states.

Add test case for suspend signal

Add test case for suspending restarting job

Add test case for HA job recovery when losing leadership

Add online documentation for the job status

uce commented on docs/internals/job_scheduling.md:
```diff
@@ -74,7 +74,28 @@ Besides the vertices, the ExecutionGraph also contains the {% gh_link /flink-run
+Once the job has been completely restarted, it will reach the *created* state.
+
+In case that the user cancels the job, it will go into the *cancelling* state.
+This is also entails the cancellation of all currently running tasks.
```
Typo: This **~~is~~** also entails...

tillrohrmann replied:
Thanks for spotting :-)

uce commented on docs/internals/job_scheduling.md:
```diff
@@ -74,7 +74,28 @@ Besides the vertices, the ExecutionGraph also contains the {% gh_link /flink-run
+Consequently, a job which reaches the *suspended* state won't be completely cleaned up.
+
+<div style="text-align: center;">
+<img src="fig/job_status.svg" alt="States and Transitions of Flink job" height="500px" style="text-align: center;"/>
```
Very nice figure! This will future contributors a lot.

uce commented on flink-runtime/src/test/resources/log4j-test.properties:
```diff
@@ -16,7 +16,7 @@
 ################################################################################
 
-log4j.rootLogger=OFF, console
+log4j.rootLogger=INFO, console
```
Should stay `OFF`?

tillrohrmann replied:
Yes, of course. Will fix it.

uce commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -1029,21 +1061,25 @@ else if (current == JobStatus.CANCELLING) {
+						break;
+					}
 					else {
 						fail(new Exception("ExecutionGraph went into final state from state " + current));
```
Missing `break` I guess (this should not be called anyways and it was that way before your change)

tillrohrmann replied:
Yes you're right. The `break` is missing here. Will add it.

uce commented:
Really good changes and great that you added the `JobStatus` graph transitions figure to the docs.

I looked at all changes and tests. Furthermore, I've tested this with lost ZooKeeper connections and it works as expected. The job is suspended and recovered smoothly. 👍 

Some general notes:
- The only thing that might be slightly confusing for users is that until the web frontend is notified about the leader changes, it shows the job status as FAILED until it disappears. This is probably some web interface quirk, because it does not recognize the new job status. I understand that you remove suspended job graphs and don't archive them (which makes sense as they are recovered later), but giving that it takes some time until all information propagates to the web frontend, we might want to temporarily show the state as suspended. We can do this as a follow up.

- Regarding the removal of restart strategy disabling, I thought about whether we should leave it in and get rid of the `SuppressRestartsException` instead. That seems to be cleaner than the custom exception, which is checked in the `ExecutionGraph`. If we want to get rid of it (as a follow up), we should leave the disable call in.

- If you think that https://github.com/apache/flink/pull/2095#issuecomment-227731879 is a valid point, then we should apply it for the suspended state as well.

These comments are not blockers and I think this is good to merge. Great job!

uce commented:
PS As another follow up we can probably get rid of some of the `JobManagerProcess` tests, which tried -- in a way -- to work around the missing suspended state. Your approach in `JobManagerHARecoveryTest` is way better.

tillrohrmann commented:
Thanks for your thorough review @uce.

- We have talked offline concerning the status in the web interface and it turned out to be not a problem since the jobs are directly removed from `currentJobs` in the `JobManager` when `cancelAndClearEverything` is called. This implies that the jobs will no longer be shown in the web interface.

- I agree that it is a good idea to replace the `SuppressRestartException` by a mechanism to disable the `RestartStrategies`. I've opened a [JIRA issue](https://issues.apache.org/jira/browse/FLINK-4112) to keep track of the effort. I think, however, that the mechanism should be implemented as part of the work on this issue. That way we won't mix pull requests with each other.

- You're right. I've addressed the comment and added `SUSPENDED` to the set of states which don't throw an exception when encountered in `ExecutionGraph.restart`.

- You're right concerning the `JobManagerProcess` tests. Maybe we can refactor some of them in the future by applying a similar pattern as it was used in `JobManagerHARecoveryTest`.

tillrohrmann commented:
Will be merging this PR.
